File: The Logic/Game/LogicGame.m
//
//  LogicGame.m
//  The Logic
//
//  Created by babi4_97 on 05.11.2017.
//  Copyright © 2017 Ivan Babkin. All rights reserved.
//

#import "LogicGame.h"
#import "ChipImageView.h"
#import <UIKit/UIKit.h>
#import "GameConst.h"

@interface LogicGame() {
    NSTimer *_timer;
}
@property (nonatomic, strong) NSArray <NSNumber *> *sequence;
@end

@implementation LogicGame

- (id)init {
    self = [super init];
    if (self) {
        NSMutableArray *arr = [[NSMutableArray alloc] init];
        for (int i = 0; i < 4; i++) {
            [arr addObject:[NSNumber numberWithInteger:arc4random() % 6]];
        }
        NSLog(@"Right answer: %@", arr);
        _time = [[Time alloc] init];
        _timer = [NSTimer scheduledTimerWithTimeInterval:1.0f target:self selector:@selector(timerTick:) userInfo:nil repeats:YES];
        _moves = 0;
        _sequence = [arr copy];
    }
    return self;
}

- (void)timerTick:(NSTimer *)timer {
    [self.time increaseBySecond];
}

- (NSArray *)checkAnswer:(NSMutableArray <NSNumber *> *)nums imgs:(NSArray *)sequenceImgs {
    self.moves++;
    NSMutableArray *answers = [[NSMutableArray alloc] init];
    NSMutableArray *sequence = [[NSMutableArray alloc] initWithArray:[self.sequence copy]];
    for (int i = 0; i < 4; i++) {
        if ([sequence[i] isEqual:nums[i]]) {
            [answers addObject:[NSNumber numberWithInt:CorrectColorAndPlace]];
            nums[i] = @(-1);
            sequence[i] = @(-1);
        }
    }
    
    for (int i = 0; i < 4; i++) {
        for (int j = 0; j < 4; j++) {
            if ([sequence[i] isEqual:nums[j]] && ![sequence[i] isEqual:@(-1)]) {
                [answers addObject:[NSNumber numberWithInt:CorrectColor]];
                nums[j] = @(-1);
                sequence[i] = @(-1);
            }
        }
    }
    
    if ([self endGame:answers]) {
        [_timer invalidate];
        _timer = nil;
        [self wrightBestMoves];
        [self wrightBestTime];
        [self.delegate endGame];
    }
    
    return answers;
}

- (NSArray *)checkAnswerOfPlayerSequence:(NSArray *)sequenceImgs {
    NSMutableArray *nums = [[NSMutableArray alloc] init];
    for (ChipImageView *img in sequenceImgs) {
        NSNumber *num = @(img.color);
        [nums addObject:num];
    }
    return [self checkAnswer:nums imgs:sequenceImgs];
}

- (bool)endGame:(NSArray *)nums {
    return [@[@2,@2,@2,@2] isEqual:nums];
}

- (void)wrightBestMoves {
    @autoreleasepool {
        NSUserDefaults *ud = [NSUserDefaults standardUserDefaults];
        NSInteger bestMoves = [ud integerForKey:kGameConstBestMoves];
        if (bestMoves == 0 || self.moves < bestMoves) {
            [ud setInteger:self.moves forKey:kGameConstBestMoves];
        }
    }
}

- (void)wrightBestTime {
    @autoreleasepool {
        NSUserDefaults *ud = [NSUserDefaults standardUserDefaults];
        Time *bestTime = (Time *)[NSKeyedUnarchiver unarchiveObjectWithData:[ud objectForKey:kGameConstBestTime]];
        
        if (bestTime) {
            if ([bestTime biggerThan:self.time]) {
                [ud setObject:[NSKeyedArchiver archivedDataWithRootObject:self.time] forKey:kGameConstBestTime];
            }
        }
        else {
            [ud setObject:[NSKeyedArchiver archivedDataWithRootObject:self.time] forKey:kGameConstBestTime];
        }
    }
}

@end

File: The logic/Game/LogicGame.swift
//
//  LogicGame.swift
//  The logic
//
//  Created by Ivan Babkin on 12.06.2018.
//  Copyright © 2018 Ivan Babkin. All rights reserved.
//

import UIKit

enum AnswerType {
    case IncorrectColor
    case CorrectColor
    case CorrectColorAndPlace
}

protocol LogicGameDelegate: class {
    func endGame()
}

class LogicGame: NSObject {
    private var timer: Timer!
    private let sequence: Array<ChipColor>
    
    var time: Time
    var moves = 0
    
    weak var delegate: LogicGameDelegate!
    
    override init() {
        var hSequence = Array<ChipColor>()
        
        for _ in 0...3 {
            hSequence.append(ChipColor.getRandomColor())
        }
        
        print(hSequence)
        
        sequence = hSequence
        time = Time()
        
        super.init()
        timer = Timer.scheduledTimer(timeInterval: 1.0, target: self, selector: #selector(timerTick), userInfo: nil, repeats: true)
    }
    
    @objc func timerTick() {
        time.increaseBySecond()
    }
    
    func checkAnswer(nums: Array<ChipColor>, sequenceImgs: Array<ChipImageView>) -> Array<AnswerType> {
        self.moves += 1
        var sequence: Array<ChipColor> =  self.sequence
        var answers = Array<AnswerType>()
        var nums = nums
        
        for i in 0...3 {
            if sequence[i] == nums[i] {
                answers.append(.CorrectColorAndPlace)
                nums[i] = .None
                sequence[i] = .None
            }
        }
        
        for i in 0...3 {
            for j in 0...3 {
                if sequence[i] == nums[j] && sequence[i] != .None {
                    answers.append(.CorrectColor)
                    nums[j] = .None
                    sequence[i] = .None
                }
            }
        }
        
        if endGame(answers: answers) {
            timer.invalidate()
            writeBestMoves()
            writeBestTime()
            delegate.endGame()
        }
        
        return answers
    }
    
    func checkAnswerOfPlayerSequence(_ sequenceImg: Array<ChipImageView>) -> Array<AnswerType> {
        var colors = Array<ChipColor>()
        for img in sequenceImg {
            colors.append(img.color)
        }
        
        return checkAnswer(nums: colors, sequenceImgs: sequenceImg)
    }
    
    private func endGame(answers: Array<AnswerType>) -> Bool {
        let a = AnswerType.CorrectColorAndPlace
        return [a, a, a, a] == answers
    }
    
    private func writeBestMoves() {
        let ud = UserDefaults.standard
        let bestMoves = ud.integer(forKey: Constants.kBestMoves)
        
        if bestMoves == 0 || self.moves < bestMoves {
            ud.set(self.moves, forKey: Constants.kBestMoves)
        }
    }
    
    private func writeBestTime() {
        let ud = UserDefaults.standard
        if let dataBestTime = ud.data(forKey: Constants.kBestTime) {
            let bestTime = NSKeyedUnarchiver.unarchiveObject(with: dataBestTime) as! Time
            if bestTime.biggerThan(time: time) {
                ud.set(NSKeyedArchiver.archivedData(withRootObject: time), forKey: Constants.kBestTime)
            }
        }
        else {
            ud.set(NSKeyedArchiver.archivedData(withRootObject: time), forKey: Constants.kBestTime)
        }
    }
}
